src/main/java/net/neoforged/neoforge/registries/DeferredRegister.java

```diff
@@ -5,319 +5,213 @@
+     * @param name The new entry's name. It will automatically have the {@linkplain #getNamespace() namespace} prefixed.
+     * @param func A factory for the new entry. The factory should not cache the created entry.
+     * @return A DeferredHolder that will track updates from the registry for this entry.
      */
-    @SuppressWarnings({ "rawtypes", "unchecked" })
-    public <I extends T> RegistryObject<I> register(final String name, final Supplier<? extends I> sup) {
+    public <I extends T> DeferredHolder<T, I> register(final String name, final Function<ResourceLocation, ? extends I> func) {
         if (seenRegisterEvent)
             throw new IllegalStateException("Cannot register new entries to DeferredRegister after RegisterEvent has been fired.");
         Objects.requireNonNull(name);
-        Objects.requireNonNull(sup);
-        final ResourceLocation key = new ResourceLocation(modid, name);
-
-        RegistryObject<I> ret;
-        if (this.registryKey != null)
-            ret = this.optionalRegistry
-                    ? RegistryObject.createOptional(key, this.registryKey, this.modid)
-                    : RegistryObject.create(key, this.registryKey, this.modid);
-        else
-            throw new IllegalStateException("Could not create RegistryObject in DeferredRegister");
-
-        if (entries.putIfAbsent((RegistryObject<T>) ret, sup) != null) {
+        Objects.requireNonNull(func);
+        final ResourceLocation key = new ResourceLocation(namespace, name);
+
+        DeferredHolder<T, I> ret = DeferredHolder.create(this.registryKey, key);
+
+        if (entries.putIfAbsent(ret, () -> func.apply(key)) != null) {
             throw new IllegalArgumentException("Duplicate registration " + name);
         }
 
         return ret;
     }
 
     /**
-     * Only used for custom registries to fill the forge registry held in this DeferredRegister.
+     * This method is used to configure a custom modded registry. It can only be invoked by a single DeferredRegister instance for a given registry key.
      *
-     * Calls {@link RegistryBuilder#setName} automatically.
-     *
-     * @param sup Supplier of a RegistryBuilder that initializes a {@link IForgeRegistry} during the {@link NewRegistryEvent} event
-     * @return A supplier of the {@link IForgeRegistry} created by the builder.
-     *         Will always return null until after the {@link NewRegistryEvent} event fires.
+     * @param consumer A consumer that configures the provided RegistryBuilder during {@link NewRegistryEvent}
+     * @return The {@link Registry} linked to {@link #getRegistryKey()}.
      */
-    public Supplier<IForgeRegistry<T>> makeRegistry(final Supplier<RegistryBuilder<T>> sup) {
-        return makeRegistry(this.registryKey.location(), sup);
+    public Registry<T> makeRegistry(final Consumer<RegistryBuilder<T>> consumer) {
+        return makeRegistry(this.registryKey.location(), consumer);
     }
```

Comment:
Registry creation should be removed from DR and replaced with direct usage of RegistryBuilder - i.e. `Registry<R> REG = RegistryBuilder.create(key).opts(...).build()`

Reply:
Indeed, I also suggested making it fully static. (And only allowed before the registry events).

Reply:
If the suggestion is to get rid of `NewRegistryEvent`, that's a bad idea. A static list (thread-safe or not) to which you just add registries-to-be-registered is as bad as any vanilla collection with no event. The entire point was to move to events for registering such constructs because it leads to class loading magic issues.

Reply:
We should get rid of the event because it doesn't serve any practical purpose anymore. You could just put the registry in a private static list somewhere, which is exactly the same as registering an event handler anyway... (the event handlers are also a private static list somewhere)